src/common/io_exerciser/DataGenerator.h

```diff
@@ -23,149 +23,139 @@
-
-          bufferptr generate_block(uint64_t offset) override;
-          bufferptr generate_wrong_block(uint64_t offset) override;
-          bool validate(bufferlist& bufferlist, uint64_t offset,
-                        uint64_t length) override;
-
-        private:
-          using UniqueIdBytes = uint64_t;
-          using SeedBytes = int;
-          using TimeBytes = uint64_t;
-
-          enum class ErrorType {
-            RUN_ID_MISMATCH,
-            SEED_MISMATCH,
-            DATA_MISMATCH,
-            DATA_NOT_FOUND,
-            UNKNOWN
-          };
-
-          constexpr uint8_t headerStart() const
-            { return 0; };
-          constexpr uint8_t uniqueIdStart() const
-            { return headerStart(); };
-          constexpr uint8_t uniqueIdLength() const
-            { return sizeof(UniqueIdBytes); };
-          constexpr uint8_t seedStart() const
-            { return uniqueIdStart() + uniqueIdLength(); };
-          constexpr uint8_t seedLength() const
-            { return sizeof(SeedBytes); };
-          constexpr uint8_t timeStart() const
-            { return seedStart() + seedLength(); };
-          constexpr uint8_t timeLength() const
-            { return sizeof(TimeBytes); };
-          constexpr uint8_t timeEnd() const
-            { return timeStart() + timeLength(); };
-          constexpr uint8_t headerLength() const
-            { return uniqueIdLength() + seedLength() + timeLength(); };
-          constexpr uint8_t bodyStart() const
-            { return headerStart() + headerLength(); };
-
-          const UniqueIdBytes readUniqueRunId(uint64_t block_offset,
-                                              const bufferlist& bufferlist);
-          const SeedBytes readSeed(uint64_t block_offset,
-                                   const bufferlist& bufferlist);
-          const TimeBytes readDateTime(uint64_t block_offset,
+namespace io_exerciser {
+namespace data_generation {
+enum class GenerationType {
+  SeededRandom,
+  HeaderedSeededRandom
+  // CompressedGenerator
+  // MixedGenerator
+};
+
+class DataGenerator {
+ public:
+  virtual ~DataGenerator() = default;
+  static std::unique_ptr<DataGenerator> create_generator(
+      GenerationType generatorType, const ObjectModel& model);
+  virtual bufferlist generate_data(uint64_t length, uint64_t offset) = 0;
+  virtual bool validate(bufferlist& bufferlist, uint64_t offset,
+                        uint64_t length);
+
+  // Used for testing debug outputs from data generation
+  virtual bufferlist generate_wrong_data(uint64_t offset, uint64_t length);
+
+ protected:
+  const ObjectModel& m_model;
+
+  DataGenerator(const ObjectModel& model) : m_model(model) {}
+};
+
+class SeededRandomGenerator : public DataGenerator {
+ public:
+  SeededRandomGenerator(const ObjectModel& model) : DataGenerator(model) {}
+
+  virtual bufferptr generate_block(uint64_t offset);
+  virtual bufferlist generate_data(uint64_t length, uint64_t offset) override;
```

Comment:
Note that we usually (always?) follow the rule of
'Explicitly annotate overrides of virtual functions or virtual destructors with exactly one of an override or (less frequently) final specifier. Do not use virtual when declaring an override. '

i.e. - do not repeat the 'virtual' specifier for overridden functions

Reply:
Sorry for missing this earlier

Reply:
No worries, good catch managing to spot it at this point! I have pushed a new commit with a fix for this